gematria/datasets/pipelines/compile_modules_lib.py

```diff
@@ -0,0 +1,156 @@
+from gematria.datasets.python import extract_bbs_from_obj
+
+
+def _get_llvm_binary_path(tool_name: str) -> str:
+  runfiles_env = runfiles.Create(os.environ)
+  assert runfiles_env is not None
+  return runfiles_env.Rlocation('llvm-project/llvm/' + tool_name)
+
+
+class OptimizeModules(beam.DoFn):
+  """A Beam function that uses LLVM opt to optimize bitcode modules."""
+
+  def __init__(self, optimization_pass_lists: list[str]):
+    self.optimization_pass_lists = optimization_pass_lists
+    self.opt_path = _get_llvm_binary_path('opt')
+
+  def optimize_module(
+      self, input_module: bytes, optimization_pass_list: list[str]
+  ) -> bytes:
+    command_vector = [self.opt_path, f'-passes={optimization_pass_list}']
+    with subprocess.Popen(
+        command_vector,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    ) as opt_process:
+      (output_bc, stderr) = opt_process.communicate(input=input_module)
+      if opt_process.returncode != 0:
+        logging.error(stderr)
+        raise ValueError('Expected opt to return 0')
+      return output_bc
+
+  def process(self, input_module: bytes) -> Iterable[bytes]:
+    for optimization_pass_list in self.optimization_pass_lists:
+      try:
+        yield self.optimize_module(input_module, optimization_pass_list)
+      except ValueError:
+        continue
+
+
+class LowerModulesAsm(beam.DoFn):
+  """A Beam function that uses LLVM llc to lower bitcode modules to object
+  files.
+  """
+
+  def __init__(self, optimization_levels: list[str]):
+    self.optimization_levels = optimization_levels
+    self.llc_path = _get_llvm_binary_path('llc')
+
+  def lower_module(self, optimization_level: str, input_module: bytes) -> bytes:
+    command_vector = [
+        self.llc_path,
+        optimization_level,
+        '-filetype=obj',
+        '-basic-block-sections=labels',
+    ]
+    with subprocess.Popen(
+        command_vector,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    ) as llc_process:
+      (output_obj, stderr) = llc_process.communicate(input=input_module)
+      if llc_process.returncode != 0:
+        logging.error(stderr)
+        raise ValueError('Expected llc to return 0')
+      return output_obj
+
+  def process(self, input_module: bytes) -> Iterable[bytes]:
+    for optimization_level in self.optimization_levels:
+      try:
+        yield self.lower_module(optimization_level, input_module)
+      except ValueError:
+        continue
+
+
+class GetBBsFromModule(beam.DoFn):
+  """A Beam function to extract BB hex values from object files."""
+
+  def process(self, input_object_file: bytes) -> Iterable[str]:
+    for bb_hex_value in extract_bbs_from_obj.get_basic_block_hex_values(
+        input_object_file
+    ):
+      yield bb_hex_value
+
+
+class DeduplicateBBs(beam.ptransform.PTransform):
+  """A Beam transform to deduplicate string data."""
+
+  def expand(self, pcoll):
```

Comment:
Can we have a more precise type annotation for `pcoll`? If not, please add a docstring explaining what's the expected structure of the pcollection.